vmware_dvs/utils/dvs_util.py

```diff
@@ -362,39 +357,40 @@ def _get_object_by_type(results, type_value):
 
         Get the desired object from the given objects result by the given type.
         """
-        return [obj for obj in results
-                if obj._type == type_value]
+        return [obj for obj in results if obj._type == type_value]
 
     def _get_ports_for_pg(self, pg_name):
         pg = self._get_pg_by_name(pg_name)
         return self.connection.invoke_api(
             vim_util, 'get_object_property',
             self.connection.vim, pg, 'portKeys')[0]
 
+    def _get_free_pg_keys(self, port_group):
+        criteria = self.builder.port_criteria(
+            port_group_key=port_group.value)
+        all_port_keys = set(
+            self.connection.invoke_api(self.connection.vim,
+                                       'FetchDVPortKeys',
+                                       self._dvs, criteria=criteria))
+        criteria.connected = True
+        connected_port_keys = set(
+            self.connection.invoke_api(self.connection.vim,
+                                       'FetchDVPortKeys',
+                                       self._dvs, criteria=criteria))
+        return list(all_port_keys - connected_port_keys - self._blocked_ports)
+
     def _lookup_unbound_port(self, port_group):
-        criteria = self.builder.port_criteria(port_group_key=port_group.value)
-        all_port_keys = self.connection.invoke_api(
-            self.connection.vim,
-            'FetchDVPortKeys',
-            self._dvs, criteria=criteria)
-        criteria = self.builder.port_criteria(port_group_key=port_group.value,
-                                              connected=True)
-        connected_port_keys = self.connection.invoke_api(
-            self.connection.vim,
-            'FetchDVPortKeys',
-            self._dvs, criteria=criteria)
-        for port_key in all_port_keys:
-            if (port_key not in connected_port_keys and
-                    port_key not in self._blocked_ports):
-                self._blocked_ports.add(port_key)
-                p_info = self._get_port_info_by_portkey(port_key)
-                if not getattr(p_info.config, 'name', None):
-                    return p_info
+        for port_key in self._get_free_pg_keys(port_group):
```

Comment:
function _get_free_pg_keys can return multiple ports (list of ports) but _lookup_unbound_port returns only one. Next time you call _get_free_pg_keys again (with 2 calls to vsphere!) instead of reuse list taken last time.